controllers/twophase/state_machine.go

```diff
@@ -0,0 +1,173 @@
+func (m *chaosStateMachine) run(ctx context.Context, targetPhase v1alpha1.ExperimentPhase) (bool, error) {
+	updated := false
+
+	duration, err := m.Chaos.GetDuration()
+	if err != nil {
+		m.Log.Error(err, "failed to get chaos duration")
+		return updated, err
+	}
+	if duration == nil {
+		zero := time.Duration(0)
+		duration = &zero
+	}
+	if m.Chaos.GetScheduler() == nil {
+		m.Log.Info("Scheduler should be defined currently")
+		return updated, fmt.Errorf("misdefined scheduler")
+	}
+
+	status := m.Chaos.GetStatus()
+	currentPhase := m.Chaos.GetStatus().Experiment.Phase
+	m.Log.Info("change phase", "current phase", currentPhase, "target phase", targetPhase)
+
+	// covers 6 conditions
+	if currentPhase == targetPhase {
+		return updated, nil
+	}
+
+	// covers 5 conditions
+	if currentPhase == v1alpha1.ExperimentPhaseFinished {
+		return updated, unexpectedInto(currentPhase, targetPhase)
+	}
+
+	// In this branch `targetPhase != v1alpha1.ExperimentPhaseRunning` is always true
+	// move from Running phase to any phase should recover the experiment.
+	// covers 5 conditions
+	if currentPhase == v1alpha1.ExperimentPhaseRunning {
+		updated = true
+
+		m.Log.Info("recovering", "current phase", currentPhase, "target phase", targetPhase)
+		err := m.Recover(ctx, m.Req, m.Chaos)
+		if err != nil {
+			status.FailedMessage = err.Error()
+
+			m.Log.Error(err, "fail to recover")
+			return updated, err
+		}
+		status.Experiment.Phase = targetPhase
+
+		now := time.Now()
+		status.Experiment.EndTime = &metav1.Time{
+			Time: now,
+		}
+		status.Experiment.Duration = now.Sub(status.Experiment.StartTime.Time).String()
+
+		// Pause from a running phase should set the nextStart to now
+		// so that the Reconciler will start another cycle of running right
+		// after resume
+		if targetPhase == v1alpha1.ExperimentPhasePaused {
+			m.Chaos.SetNextStart(now)
+		}
+	}
+
+	// currentPhase could be "" or Waiting or Paused or Failed
+	// covers 4 * 5 = 20 conditions
+	if currentPhase != v1alpha1.ExperimentPhaseRunning {
+		if targetPhase == v1alpha1.ExperimentPhaseRunning {
+			updated = true
+
+			m.Log.Info("applying", "current phase", currentPhase, "target phase", targetPhase)
+			err := m.Apply(ctx, m.Req, m.Chaos)
+			if err != nil {
+				m.Log.Error(err, "fail to apply")
+
+				status.Experiment.Phase = v1alpha1.ExperimentPhaseFailed
+				status.FailedMessage = err.Error()
+
+				return updated, err
+			}
+			status.Experiment.Phase = targetPhase
+
+			startTime := time.Now()
+			nextStart, err := utils.NextTime(*m.Chaos.GetScheduler(), startTime)
+			if err != nil {
+				m.Log.Error(err, "failed to get the next start time")
+				return updated, err
+			}
+			nextRecover := startTime.Add(*duration)
+			// if the duration is too long, `nextRecover` could be after `nextStart`
+			// we can jump over a start to make sure `nextRecover` is before `nextStart`
+			for nextRecover.After(*nextStart) {
```

Comment:
? It will only loop 60 times and get a correct time. How could it block the goroutine?

Reply:
Yes, my fault, this example will loop 60 times. If the duration is too long, such as 1 year, and the `corn` is very short, such as 10s. this function will be blocked very long.

Reply:
@cwen0 I can not even binary search here, as the scheduler could only iterate once. Do you have any suggestion?

Reply:
Could we directly set `nextStart, err = utils.NextTime(*m.Chaos.GetScheduler(), nextRecover)`, since the condition fulfill the current one would make nextStart the first start time after nextRecover